src/multiplex/client.rs

```diff
@@ -328,16 +431,16 @@ where
                     // keeping a HashMap around, and is _usually_ fast as long as the requests
                     // that have been pending the longest are most likely to complete next.
                     let id = transport.as_mut().finish_tag(&r);
-                    let pending = this
-                        .responses
-                        .iter()
-                        .position(|&Pending { ref tag, .. }| tag == &id)
-                        .ok_or(Error::Desynchronized)?;
-
-                    // this request just finished, which means it's _probably_ near the front
-                    // (i.e., was issued a while ago). so, for the swap needed for efficient
-                    // remove, we want to swap with something else that is close to the front.
-                    let pending = this.responses.swap_remove_front(pending).unwrap();
+                    let pending = match pending.as_mut().take(id) {
+                        Some(pending) => pending?,
+                        None => {
+                            tracing::trace!(
+                                "response arrived but no associated pending tag; ignoring response"
+                            );
+                            continue;
+                        }
```

Comment:
This is to support the case where it doesn't crash? The `?` is in the above branch which would terminate the future.

Reply:
Yep, I get that — I'm just saying we need to document the default behavior and how to customize it well, since reading this code the only way you'd know that the default implementation doesn't ever hit this `None` case is to go look at its `impl PendingStore` as well.

Reply:
Right, I think docs on the trait method that produces this branch is good enough. The default impl (VecDeque) doesn't hit this case but its possible a custom impl does. So I think just explaining the contract in the docs works.